### PR TITLE
test for java:solrqueryaggregator node to query NRT replica.

### PR DESCRIPTION
**Pull request by noblepaul**
https://app.shortcut.com/fullstory/story/201331/add-test-for-java-solrqueryaggregator-node-to-query-nrt-replica#activity-209224

### CODE REVIEW

**Review by magibney — COMMENTED** [review]
This is quite nice! I enhanced this a little bit around the edges (running jetty manipulations in a separate thread, e.g.); wondering if you'd be ok with me pushing to this branch? It'd probably be more straightforward to just push my modifications _somewhere_, and easier to track changes if collaborative development happens on the same branch. That said, I'm happy to push to a separate branch if you prefer.

**Comment by noblepaul** [comment]
Go ahead and commit to the branch

**Comment by magibney** [comment]
Thanks Noble! I just pushed a commit to this branch that makes a few adjustments. Feel free to revert/modify as you see fit. The goal was to make some assumptions about behavior a little more explicit, and provide a baseline for detecting performance regressions, etc.

It's still one big test, which is perhaps not ideal, but I think it's nonetheless helpful in its current state.

File: solr/core/src/test/org/apache/solr/core/FSHATest.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        // accounts for the 3000ms tolerance allowed for below. This is not a strict value, and if it causes failures
+        // regularly we should feel free to increase the tolerance; but it's meant to provide a stable baseline from
+        // which to detect regressions.
+        assertWithinTolerance(nrtDowntimeMs, now - start, 3000);
+        count = 0;
+        start = System.currentTimeMillis();
+        individualRequestStart = start;
+        while (pullCore.equals(hostCore = getHostCoreName(COLL, qaJettyBase, cluster.getSolrClient(), p -> {
+          p.set(CommonParams.Q, "id:345");
+          p.add("shards.preference", "replica.type:NRT");
```
Comment (hiteshk25) [inline solr/core/src/test/org/apache/solr/core/FSHATest.java:199]: it's optional to specify "pull" replica here ? 

CCing @magibney @noblepaul

Comment (magibney) [reply solr/core/src/test/org/apache/solr/core/FSHATest.java:199]: I'm not sure I understand the question. You could indeed specify `shards.preference=replica.type:PULL`, but this test is specifically exercising our expected use-case, where in the request we'll _always_ be asserting a preference for NRT replicas. We're just testing assumptions about what happens when our preference cannot be satisfied (i.e., the request should fallback to PULL replicas).

Comment (hiteshk25) [reply solr/core/src/test/org/apache/solr/core/FSHATest.java:199]: Thanks. Just wanted to verify this 
```
(i.e., the request should fallback to PULL replicas).
```

We pass both the preference in following way. Tested locally
```
"shards.preference", "replica.type:NRT,replica.type:PULL")
```

**Review by hiteshk25 — APPROVED** [review]
LGTM

File: solr/core/src/test/org/apache/solr/core/FSHATest.java (hunk 1 above)
Comment (magibney) [reply solr/core/src/test/org/apache/solr/core/FSHATest.java:199]: ohhh yes! I see what you mean. Those should be functionally equivalent. Sorry I misunderstood the question!

**Comment by magibney** [comment]
Jotting this down to have it somewhere; seeing this error (just extra noise, but would be good to fix by adding `[core]` doc transformer to this test):
```
   [junit4] FAILURE 0.19s J2 | TestRandomFlRTGCloud.testCoverage <<<
   [junit4]    > Throwable #1: java.lang.AssertionError: Some implicit TransformerFactories are not yet tested by this class: [core] expected:<0> but was:<1
>
   [junit4]    >        at __randomizedtesting.SeedInfo.seed([FB5DBA93EE4B3FF9:A8770A24E2459933]:0)
   [junit4]    >        at org.apache.solr.cloud.TestRandomFlRTGCloud.testCoverage(TestRandomFlRTGCloud.java:206)
   [junit4]    >        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   [junit4]    >        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   [junit4]    >        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   [junit4]    >        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   [junit4]    >        at java.base/java.lang.Thread.run(Thread.java:829)

```

**Pull request merged**